src/dune_rules/artifacts.ml

```diff
@@ -41,9 +46,13 @@ module Bin = struct
 
   let add_binaries t ~dir l =
     let local_bins =
-      List.fold_left l ~init:t.local_bins ~f:(fun acc fb ->
-          let path = File_binding.Expanded.dst_path fb ~dir:(local_bin dir) in
-          String.Map.set acc (Path.Build.basename path) path)
+      Memo.lazy_ (fun () ->
```

Comment:
It's good to at least set the name for lazy nodes. Otherwise, debugging becomes a nightmare.